File: Computer Programming/C++/Sliding Window/Find All Anagrams in a String.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) 
    {
        unordered_map<char,int>mps;
        unordered_map<char,int>mpp;
        
        for(auto it : p)
        {
            mpp[it] ++;
        }

        int left = 0;
        int right = 0;
        vector<int>ans;
        for(int right = 0;right<s.size();right++)
        {
            mps[s[right]] ++;
            if(right - left + 1 > p.size())
            {
                mps[s[left]] --;
                if(mps[s[left]] == 0)
                {
                    mps.erase(s[left]);  
                }    
                left ++;
            }

            if(right - left + 1 == p.size())
            {
                if(mpp == mps)
                {
                    ans.push_back(left);
                }
            }
        }

        return ans;
    }
};

File: Computer Programming/Python/Sliding Window/Find All Anagrams in a String.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        
        mpl = {}
        l = len(p)
        n = len(s)
        for ele in p:
            if ele in mpl:
                mpl[ele] += 1
            else:
                mpl[ele] = 1

        right = 0
        left = 0
        ans = []
        mps = {}
        for right in range(0,n):
            if s[right] in mps:
                mps[s[right]] += 1
            else:
                mps[s[right]] = 1

            if right - left + 1 == l:
                if mpl == mps:
                    ans.append(left)
                mps[s[left]] -= 1
                if mps[s[left]] == 0:
                    del mps[s[left]]
                left += 1



        return ans
